operators/advection_operator.h

```diff
@@ -0,0 +1,436 @@
+                          std::sin(numbers::PI * p[1]));
+      }
+      else if (component == 2)
+      {
+        return -factor * (std::sin(2 * numbers::PI * p[0]) *
+                          std::sin(2 * numbers::PI * p[1]) *
+                          std::sin(numbers::PI * p[2]) *
+                          std::sin(numbers::PI * p[2]));
+      }
+
+      break;
+    }
+
+    case 7:
+    {
+      // Hagen-Poiseuille flow test case (channel flow), (domain: [0,1]x[0,1])
+      double const x0 = 0.5;
+      double const y0 = 0.5;
+
+      double const R = 0.15;
+      double const nu = 0.01;
+      double const dpdx = 0.5;
+
+      if (component == 0)
+      {
+        return 0.0;
+      }
+
+      if (component == 1)
+      {
+        return 0.0;
+      }
+
+      if (component == 2)
+      {
+        double const r2 = (p[0] - x0) * (p[0] - x0) + (p[1] - y0) * (p[1] - y0);
+
+        if (r2 <= R * R)
+        {
+          return -R * R * dpdx / (4.0 * nu) * (1.0 - r2 / R / R);
+        }
+        else
+        {
+          return 0.0;
+        }
+
+        break;
+      }
+    }
+
+    default:
+      AssertThrow(false, ExcNotImplemented());
+    }
+  }
+
+  return 0;
+}
+using Number = double;
+template <int dim, int fe_degree>
+class AdvectionOperator : public OperatorBase<dim, fe_degree>
+{
+public:
+  using Number = double;
+
+  AdvectionOperator(OperatorData<dim, fe_degree> &operator_data, LevelSetProblemParameters const &param) : OperatorBase<dim, fe_degree>(operator_data), param_(param)
+  {
+  }
+
+  void reinit();
+  void set_velocity_operator(double const time) const
+  {
+
+    TransportSpeed<dim> transport_speed(time, param_);
+
+    VectorTools::interpolate(this->operator_data_.dof_handler_vel_,
+                             transport_speed,
+                             velocity_operator_);
+
+    velocity_operator_.update_ghost_values();
+  }
+
+  void apply_operator(double const time, LinearAlgebra::distributed::Vector<Number> &dst, LinearAlgebra::distributed::Vector<Number> const &src) const;
+
+  void
+  compute_RI_distance(const double vertex_distance)
+  {
+    // Compute the reinitialization distance
+    RI_distance = vertex_distance / fe_degree * param_.factor_RI_distance;
+  }
+
+      private :
+
+      mutable LinearAlgebra::distributed::Vector<Number>
+          velocity_operator_;
+  LevelSetProblemParameters const &param_;
+
+      void
+      local_apply_domain(
+          const MatrixFree<dim, Number> &data,
+          LinearAlgebra::distributed::Vector<Number> &dst,
+          const LinearAlgebra::distributed::Vector<Number> &src,
+          const std::pair<unsigned int, unsigned int> &cell_range) const;
+
+  void
+  local_apply_inner_face(
+      const MatrixFree<dim, Number> &data,
+      LinearAlgebra::distributed::Vector<Number> &dst,
+      const LinearAlgebra::distributed::Vector<Number> &src,
+      const std::pair<unsigned int, unsigned int> &cell_range) const;
+
+  void
+  local_apply_boundary_face(
```

Comment:
Based on an additional template parameter, e.g. an `enum BoundaryEvaluationType` the behavior of the loop can be controlled.

```suggestion
  template<BoundaryEvaluationType type>
  void
  local_apply_boundary_face(
```